Lexer for the WebGPU Shading Language

dneto0 opened the pull request:
See https://w3.org/TR/WGSL

Further work is needed to refine it:
- treat context-dependent names specially
- treat template start and template end tokens specially, perhaps

Anteru commented:
Hey, long time no see. I've run the checks, let's see how they go. Maybe we can squeeze this into the 2.15 release :)

dneto0 commented on pygments/lexers/wgsl.py:
```diff
@@ -0,0 +1,415 @@
+# https://www.w3.org/TR/WGSL/#syntax-ident_pattern_token
+ident_pattern_token = '([_{}][{}]+)|[{}]'.format(uni.xid_start,uni.xid_continue,uni.xid_start)
+
+def qw(s):
+    """Returns the non-whitespace tokens in a possibly multiline string.
+    Emulate Perl's qw syntax.
+    """
+    result = []
+    for line in s.splitlines():
+      for word in line.split(' '):
+        if len(word) > 0:
+          result.append(word)
+    return result
```
> What's the rationale for using this over a list of words directly? 

It's much easier to copy-paste from the spec or another source, without having to worry about details of getting the quoting just right.  Working with qw is really fluent.

> Also, this could be simplified to use `yield word` instead of appending to a result, which shouldn't matter for the `words` function

I am but a simple Python programmer.  :-)

dneto0 replied:
Indeed!  I misread the docs. Will fix.

dneto0 replied:
Fixed this. I've removed the qw function. Thanks!

dneto0 commented:
I have also added tests for the various predeclared names in WGSL, which map to 'Name.Builtin'.
Earlier I had a typo in one of the texel formats, which I already fixed.

Anteru commented:
@jeanas Anything from your side? Looks good to go for me.

dneto0 commented:
Thank you!
